drivers/mipi_dbi/mipi_dbi_stm32_fmc.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2024 Bootlin
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT st_stm32_fmc_mipi_dbi
+
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/mipi_dbi.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mipi_dbi_stm32_fmc, CONFIG_MIPI_DBI_LOG_LEVEL);
+
+#define FMC_DATA_ADDR_OFFSET 0x02
+
+struct mipi_dbi_stm32_fmc_config {
+	/* Reset GPIO */
+	const struct gpio_dt_spec reset;
+	/* Power GPIO */
+	const struct gpio_dt_spec power;
+	uint16_t *register_addr;
+	uint16_t *data_addr;
+};
+
+struct mipi_dbi_stm32_fmc_data {
+};
+
+int mipi_dbi_stm32_fmc_command_write(const struct device *dev,
+				     const struct mipi_dbi_config *dbi_config, uint8_t cmd,
+				     const uint8_t *data_buf, size_t len)
+{
+	const struct mipi_dbi_stm32_fmc_config *config = dev->config;
+
+	uint32_t i = 0;
+
+	*config->register_addr = cmd;
+
```

Comment:
The `__DSB();` might be required here and in places where `data_addr` or `register_addr` are written.

Reply:
Based on my trials with FMC the `volatile uint16_t *data_addr` doesn't help. The `__DSB` or at least `__DMB` does.

Reply:
I've added the function `barrier_dsync_fence_full` after every register access, is that okay ?

Reply:
@miggazElquez yes, thank you.

We can improve it further:

Using existing `sys_io` and `byteorder` functions.

Remove `FMC_SUBBANK_SIZE`, the FMC headers tend to define all FMC bank addresses.
Therefore, we can use `_CONCAT` and directly create required FMC bank define.
Note: it won't work correctly if the bank is remapped and I didn't include a check for `0`.
Ideally, we have to provide a function in the FMC to get the bank address.

```diff
diff --git a/d.c b/d.c
index 9eff43b..04f8485 100644
--- a/d.c
+++ b/d.c
@@ -10,6 +10,7 @@
 #include <zephyr/drivers/mipi_dbi.h>
 #include <zephyr/drivers/clock_control/stm32_clock_control.h>
 #include <zephyr/sys/barrier.h>
+#include <zephyr/sys/sys_io.h>
 
 #include <zephyr/logging/log.h>
 LOG_MODULE_REGISTER(mipi_dbi_stm32_fmc, CONFIG_MIPI_DBI_LOG_LEVEL);
@@ -21,8 +22,8 @@ struct mipi_dbi_stm32_fmc_config {
 	const struct gpio_dt_spec reset;
 	/* Power GPIO */
 	const struct gpio_dt_spec power;
-	uint16_t *register_addr;
-	uint16_t *data_addr;
+	mem_addr_t register_addr;
+	mem_addr_t data_addr;
 	uint32_t fmc_AddressSetupTime;
 	uint32_t fmc_DataSetupTime;
 	uint32_t fmc_MemoryWidth;
@@ -81,15 +82,12 @@ int mipi_dbi_stm32_fmc_command_write(const struct device *dev,
 		return ret;
 	}
 
-	uint32_t i = 0;
-
-	*config->register_addr = cmd;
+	sys_write16(cmd, config->register_addr);
 	barrier_dsync_fence_full();
-	while (i < len) {
-		uint16_t data = (uint16_t)data_buf[i];
-		*config->data_addr = data;
+
+	for (size_t i = 0U; i < len; ++i) {
+		sys_write16((uint16_t)data_buf[i], config->data_addr);
 		barrier_dsync_fence_full();
-		i += 1;
 	}
 
 	return 0;
@@ -102,7 +100,6 @@ static int mipi_dbi_stm32_fmc_write_display(const struct device *dev,
 					    enum display_pixel_format pixfmt)
 {
 	const struct mipi_dbi_stm32_fmc_config *config = dev->config;
-	uint32_t i = 0;
 	int ret;
 
 	ret = mipi_dbi_stm32_fmc_check_config(dev, dbi_config);
@@ -110,11 +107,9 @@ static int mipi_dbi_stm32_fmc_write_display(const struct device *dev,
 		return ret;
 	}
 
-	while (i < desc->buf_size) {
-		uint16_t data = ((uint16_t)framebuf[i + 1U] << 8U) | (uint16_t)framebuf[i];
-		*config->data_addr = data;
+	for (size_t i = 0U; i < desc->buf_size; i += 2U) {
+d		sys_write16(sys_get_le16(&framebuf[i]), config->data_addr);
 		barrier_dsync_fence_full();
-		i += 2;
 	}
 
 	return 0;
@@ -135,6 +130,7 @@ static int mipi_dbi_stm32_fmc_reset(const struct device *dev, uint32_t delay)
 	}
 
 	k_msleep(delay);
+
 	return gpio_pin_set_dt(&config->reset, 0);
 }
 
@@ -175,17 +171,14 @@ static struct mipi_dbi_driver_api mipi_dbi_stm32_fmc_driver_api = {
 	.write_display = mipi_dbi_stm32_fmc_write_display,
 };
 
-/* Offset between the address of the subbanks of the bank 1 */
-#define FMC_SUBBANK_SIZE 0x04000000UL
-
-#define MIPI_DBI_FMC_GET_ADDRESS(n) FMC_BANK1 + DT_REG_ADDR(DT_INST_PARENT(n)) * FMC_SUBBANK_SIZE
+#define MIPI_DBI_FMC_GET_ADDRESS(n) _CONCAT(FMC_BANK, DT_REG_ADDR(DT_INST_PARENT(n)));
 
 #define MIPI_DBI_STM32_FMC_INIT(n)                                                                 \
 	static const struct mipi_dbi_stm32_fmc_config mipi_dbi_stm32_fmc_config_##n = {            \
 		.reset = GPIO_DT_SPEC_INST_GET_OR(n, reset_gpios, {}),                             \
 		.power = GPIO_DT_SPEC_INST_GET_OR(n, power_gpios, {}),                             \
-		.register_addr = (uint16_t *)MIPI_DBI_FMC_GET_ADDRESS(n),                          \
-		.data_addr = (uint16_t *)(MIPI_DBI_FMC_GET_ADDRESS(n) + FMC_DATA_ADDR_OFFSET),     \
+		.register_addr = MIPI_DBI_FMC_GET_ADDRESS(n),                                      \
+		.data_addr = MIPI_DBI_FMC_GET_ADDRESS(n) + FMC_DATA_ADDR_OFFSET,                   \
 		.fmc_AddressSetupTime = DT_PROP_BY_IDX(DT_INST_PARENT(n), st_timing, 0),           \
 		.fmc_DataSetupTime = DT_PROP_BY_IDX(DT_INST_PARENT(n), st_timing, 2),              \
 		.fmc_MemoryWidth = DT_PROP_BY_IDX(DT_INST_PARENT(n), st_control, 2),               \

```

Reply:
I have two question : if we use `sys_write16`, do we still need to call `barrier_dsync_fence_full` ? It seems like the function already call `barrier_dmem_fence_full`.

For the FMC bank, the problem is that in the device tree, the adress is 0 to 3, but in the header it's 1 to 4. I think `_CONCAT(FMC_BANK1_, DT_REG_ADDR(DT_INST_PARENT(n))+1);` wouldn't work ?

Reply:
>  do we still need to call barrier_dsync_fence_full ? It seems like the function already call barrier_dmem_fence_full.

That depends on the target. I see that only cortex-a-r and arm64 include the barrier.

> For the FMC bank, the problem is that in the device tree, the adress is 0 to 3, but in the header it's 1 to 4. I think _CONCAT(FMC_BANK1_, DT_REG_ADDR(DT_INST_PARENT(n))+1); wouldn't work ?

Then, `_CONCAT(FMC_BANK1_, UTIL_INC(DT_REG_ADDR(DT_INST_PARENT(n))); `

Reply:
Maybe it is worth to add barriers on/off Kconfig option knowing that some MCUs don't need the barrier (_to get some performance boost_)?

Could be something like:

```c
config MIPI_DBI_STM32_FMC_MEM_BARRIER
    bool "Adds memory barrier after every address and data register access"
```

Then data and address register access the driver could be factored out to `ALWAYS_INLINE` function with included:
```c
if (IS_ENABLED(CONFIG_MIPI_DBI_STM32_FMC_MEM_BARRIER)) {
    barrier_dsync_fence_full();
}
```

Reply:
> Then, `_CONCAT(FMC_BANK, UTIL_INC(DT_REG_ADDR(DT_INST_PARENT(n))); `

Oh I didn't know it existed, thanks.

> Maybe it is worth to add barriers on/off Kconfig option knowing that some MCUs don't need the barrier (to get some performance boost)?

So the idea would be to not have the barrier if the barrier is already there in the call to `sys_write16` ? I understand the idea, but wouldn't it be simpler to just not use `sys_write16` then ? Or is there another advantage to use it ? Cleaner code maybe ?

Reply:
> So the idea would be to not have the barrier if the barrier is already there in the call to `sys_write16`?

The idea is to allow people to enable/disable the barrier depending on their needs. If I am not mistaken, it works for you on stm32l562 without the barriers (as they were not in the initial PR). However, stm32h7 fails without them.

> Or is there another advantage to use it? Cleaner code maybe?

Yes.

Reply:
I've done the modifications. I put the barriers on by default, I suppose I should disable them on my MCU by default, as they are not useful there (it seems like) ?

Reply:
Thanks, you can add
```

if MIPI_DBI_STM32_FMC
	config MIPI_DBI_STM32_FMC_MEM_BARRIER
	default n
endif
```
to `stm32l562e_dk/Kconfig.defconfig` to get some performance (_if it works fine without the barrier_).